Vite: Fix issues when loading files via static asset queries

philipp-spiess opened the pull request:
Fixes: #14558

This PR fixes an issue where our Vite plugin would crash when trying to load stylesheets via certain static asset query parameters:

```ts
import raw from './style.css?raw'
import url from './style.css?url'
```

The proper behavior for our extension is to _not touch these file at all_. The `?raw` identifier should never transform anything and the `?url` one will emit a module which points to the asset URL. However, if that URL is loaded as a stylesheet, another transform hook is called and the file is properly transformed. I verified this in the Vite setup and have added an integration test ensuring these two features work as expected.

I've also greatly reduced the complexity of the Vite playground to make it easier to set up examples like this in the future.

philipp-spiess commented:
* **#14716** <a href="https://app.graphite.dev/github/pr/tailwindlabs/tailwindcss/14716?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `next`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @philipp-spiess and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

adamwathan commented on packages/@tailwindcss-vite/src/index.ts:
```diff
@@ -261,9 +261,13 @@ function getExtension(id: string) {
+      (extension === 'vue' && id.includes('&lang.css')) ||
+      (extension === 'astro' && id.includes('&lang.css'))) &&
+    // Don't intercept ?url and ?raw static asset handling
+    !id.includes('?url') &&
+    !id.includes('?raw')
```
Is there ever a case where these are not at the end of the string? If not maybe `endsWith` is clearer?

bitofbreeze replied:
or could combine into a single check with some regex like `.test(/\?(url|raw)$/)`

aaronadamsCA replied:
This seems to be the regex Vite uses:

https://github.com/vitejs/vite/blob/257883fccc36f27cc879af249eeba353ebe94cdd/packages/vite/src/node/constants.ts#L86

(Sorry, stumbled across this while searching for something unrelated)

philipp-spiess replied:
@adamwathan yeah I think it can be anywhere in the query string. Doesn't even have to be first place so we also need to look for `&url` or `&raw`. Going with a modified version of the regex instead ty @aaronadamsCA (we don't have to worry about workers/shared workers though)

philipp-spiess commented on packages/@tailwindcss-vite/src/index.ts:
```diff
@@ -261,9 +263,12 @@ function getExtension(id: string) {
+      (extension === 'vue' && id.includes('&lang.css')) ||
+      (extension === 'astro' && id.includes('&lang.css'))) &&
+    // Don't intercept special static asset resources
+    !id.match(SPECIAL_QUERY_RE)
```
```suggestion
    !SPECIAL_QUERY_RE.test(id)
```

philipp-spiess committed:
Vite: Fix issues when loading files via static asset queries (#14716)

Fixes: #14558

This PR fixes an issue where our Vite plugin would crash when trying to load stylesheets via certain static asset query parameters:

```ts
import raw from './style.css?raw'
import url from './style.css?url'
```

The proper behavior for our extension is to _not touch these file at all_. The `?raw` identifier should never transform anything and the `?url` one will emit a module which points to the asset URL. However, if that URL is loaded as a stylesheet, another transform hook is called and the file is properly transformed. I verified this in the Vite setup and have added an integration test ensuring these two features work as expected.

I've also greatly reduced the complexity of the Vite playground to make it easier to set up examples like this in the future.
